File: ts/earthbucks-lib/src/ecdsa.ts
import { sign, verify } from "@earthbucks/secp256k1";
import { blake3_mac } from "@earthbucks/blake3";
import { PrivKey } from "./priv-key.js";
import { PubKey } from "./pub-key.js";
import { FixedBuf, SysBuf } from "./buf.js";

export function ecdsa_sign(
  hashBuf: FixedBuf<32>,
  privKey: PrivKey,
): FixedBuf<64> {
  const k = blake3_mac(privKey.toBuf().buf, hashBuf.buf);
  const sig = sign(hashBuf.buf, privKey.toBuf().buf, k);
  return FixedBuf.fromBuf(64, SysBuf.from(sig));
}

export function ecdsa_verify(
  sig: FixedBuf<64>,
  hashBuf: FixedBuf<32>,
  pubKey: PubKey,
): boolean {
  try {
    verify(sig.buf, hashBuf.buf, pubKey.toBuf().buf);
  } catch (e) {
    return false;
  }
  return true;
}

File: rs/earthbucks_lib/src/ecdsa.rs
use crate::hash::blake3_mac;
use earthbucks_secp256k1::secp256k1;
use crate::priv_key::PrivKey;
use crate::pub_key::PubKey;

pub fn sign(data: &[u8], priv_key: PrivKey) -> [u8; 64] {
    let key = priv_key.buf;
    let k = blake3_mac(&key, data);
    let sig = secp256k1::sign(data, &key, &k)
        .map_err(|_| "Failed to sign")
        .unwrap();
    sig.try_into().unwrap()
}

pub fn verify(data: &[u8], sig: &[u8; 64], pubkey: &[u8; 33]) -> bool {
    let res = secp256k1::verify(data, sig, pubkey);
    res.is_ok()
}
